src/froll.c

```diff
@@ -406,11 +405,253 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
+      }
+      if (verbose)
+        snprintf(end(ans->message[0]), 500, _("%s: NA (or other non-finite) value(s) are present in input, skip non-NA attempt and run with extra care for NAs\n"), __func__);
+      w = R_NegInf;
+      cmax = 0;
+    }
+  }
+  if (truehasna) {
+    int nc = 0;
+    int i;
+    for (i=0; i<k-1; i++) { // #loop_counter_not_local_scope_ok
+      if (R_FINITE(x[i])) {
+        if (x[i] >= w) {
+          iw = i;
+          w = x[iw];
+        }
+      } else {
+        nc++;
+      }
+      ans->dbl_v[i] = fill;
+    }
+    if (R_FINITE(x[i])) {
+      if (x[i] >= w) {
+        iw = i;
+        w = x[iw];
+      }
+    } else {
+      nc++;
+    }
+    if (nc == 0) {
+      ans->dbl_v[i] = w;
+    } else if (nc == k) {
+      ans->dbl_v[i] = narm ? R_NegInf : NA_REAL;
+    } else {
+      ans->dbl_v[i] = narm ? w : NA_REAL;
+    }
+    for (uint64_t i=k; i<nx; i++) {
+      if (R_FINITE(x[i])) {
+        if (x[i] >= w) {
+          iw = i;
+          w = x[iw];
+        } else {
+          if (iw > i-k) { // max is still within window
+            // w and iw are still ok
+          } else { // max left the window, need to find max in this window
+            w = R_NegInf;
+            iw = i-k;
+            windowmaxnarm(x, i, k, narm, &nc, &w, &iw);
+            //Rprintf("iteration %d, windowmaxnarm found new max %f at %d\n", i, w, iw);
+            cmax++;
+          }
+        }
+        ans->dbl_v[i] = w;
+      } else {
+        nc++;
+      }
+      if (!R_FINITE(x[i-k])) {
+        nc--;
+      }
+      if (nc == 0) {
+        ans->dbl_v[i] = w;
+      } else if (nc == k) {
+        ans->dbl_v[i] = narm ? R_NegInf : NA_REAL;
+      } else {
+        ans->dbl_v[i] = narm ? w : NA_REAL;
+      }
+    }
+    if (verbose)
+      snprintf(end(ans->message[0]), 500, _("%s: windowmaxnarm called %"PRIu64" time(s)\n"), __func__, cmax);
+  }
+}
+/* fast rolling max - exact
+ * for each observation in x compute max in window from scratch
+ * no proper support for NAs yet
+ */
+void frollmaxExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasna, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running in parallel for input length %"PRIu64", window %d, hasna %d, narm %d\n"), "frollmaxExact", (uint64_t)nx, k, hasna, (int)narm);
+  for (int i=0; i<k-1; i++) {
+    ans->dbl_v[i] = fill;
+  }
+  bool truehasna = hasna>0;
+  if (!truehasna || !narm) {
+    #pragma omp parallel for num_threads(getDTthreads(nx, true))
+    for (uint64_t i=k-1; i<nx; i++) {
+      if (narm && truehasna) {
+        continue;
+      }
+      double w = R_NegInf;
+      for (int j=-k+1; j<=0; j++) {
+        if (x[i+j] > w)
+          w = x[i+j];
+      }
+      if (R_FINITE(w)) {
+        ans->dbl_v[i] = w;
+      } else {
+        if (!narm) {
+          ans->dbl_v[i] = w;
+        }
+        truehasna = true;
```

Comment:
truehasna is shared. Should we write with `#pragma omp atomic write`?

Reply:
It is shared but it can only be updated to `true`, it is a flag to be used to stop the loop (by stop it actually skips all further iterations). Any thread that will update it to `true` will just tell to all threads to skip further iterations. If all threads at once will write to this variable, they will write `true` so behavior will be correct. atomic write would be necessary if we could also write `false` there.